File: java/0271-encode-and-decode-strings/Solution.java
import java.util.ArrayList;
import java.util.List;

public class Solution {
    public String encode(List<String> strs) {
        StringBuilder result = new StringBuilder();

        for (String s : strs) {
            result.append(Integer.toString(s.length())).append(' ').append(s);
        }

        return result.toString();
    }

    public List<String> decode(String s) {
        List<String> result = new ArrayList<>();

        int i = 0;
        while (i < s.length()) {
            int j = i;
            while (s.charAt(j) != ' ') {
                ++j;
            }

            int length = Integer.parseInt(s.substring(i, j));
            i = j + 1 + length;
            result.add(s.substring(j + 1, i));
        }

        return result;
    }
}

File: c++/0271-encode-and-decode-strings/solution.cpp
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    string encode(vector<string>& strs) {
        string result = "";

        for (string s : strs) {
            result += to_string(s.length()) + ' ' + s;
        }

        return result;
    }

    vector<string> decode(string s) {
        vector<string> result;

        int i = 0;
        while (i < s.length()) {
            int j = i;
            while (s[j] != ' ') {
                ++j;
            }

            int length = stoi(s.substr(i, j - i));
            string str = s.substr(j + 1, length);
            result.push_back(str);

            i = j + 1 + length;
        }

        return result;
    }
};
